src/pkg/k8s/common.go

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package k8s provides a client for interacting with a Kubernetes cluster.
+package k8s
+
+import (
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/go-logr/logr/funcr"
+	"k8s.io/client-go/kubernetes"
+
+	// Include the cloud auth plugins
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// New creates a new K8s client.
+func New(logger Log, defaultLabels Labels) (*K8s, error) {
+	logger("k8s.NewK8sClient()")
+
+	klog.SetLogger(funcr.New(func(prefix, args string) {
+		logger(args)
+	}, funcr.Options{}))
+
+	config, clientset, err := connect()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to k8s cluster: %w", err)
+	}
+
+	return &K8s{
+		RestConfig: config,
+		Clientset:  clientset,
+		Log:        logger,
+		Labels:     defaultLabels,
+	}, nil
+}
+
+// NewWithWait is a convenience function that creates a new K8s client and waits for the cluster to be healthy.
+func NewWithWait(logger Log, defaultLabels Labels, timeout time.Duration) (*K8s, error) {
+	k, _ := New(logger, defaultLabels)
```

Comment:
We probably shouldn't be dogsledding this in the first place. `New()` will return an error if we can't connect to the cluster at all. The wait that happens later in this function is so that we wait for at least one pod in the cluster (that we already established we can connect to) to return `succeeded` or `running`.